### PR TITLE
[KVCache] INT8 KV cache implementation and related changes

### PR DESCRIPTION
**Pull request by pujiang2018**
Main changes:
1. INT8 KV cache implementation
2. Move crossAttnShardedHead from layers to kernels

### CODE REVIEW

File: src/layers/attention.h [hunk 1]
```diff
@@ -463,8 +463,9 @@ class Attention {
                     auto srcV = value.Row(b * ctx->inputSeqLen + seq) + h * headSize;
                     auto dstV = presentValue.getSequence(pastSeqLen + seq, b, h);
 
-                    xft::copy(dstK, srcK, headSize);
-                    xft::copy(dstV, srcV, headSize);
+                    // Suppose dstK and dstV are the same data type
```
Comment (abenmao) [inline src/layers/attention.h:466]: Since there are type conversions in storeKVCache. The old comments can be removed.

**Comment by pujiang2018** [comment]
> If it is convenient, maybe you can temporarily enable int8_t in type_selector.h to run ci test

currently no interface. should be more convenient after adding interface.

**Review by abenmao — APPROVED** [review]

File: src/layers/attention.h [hunk 2]
```diff
@@ -478,19 +478,26 @@ class Attention {
...
+            xft::storeKVCache(dst, src, ctx->attHeadSize);
         }
     }
 
     // query: M * headSize, key: N * headSize, score: M * N
-    // ldq: leading dimension of query; ldk: leading dimension of key; lds: LD of score
+    // ldq: leading dimension of query; lds: LD of score
     template <typename T1, typename T2, typename T3>
-    void gemm1(T1 *query, T2 *key, T3 *score, int M, int N, int headSize, int ldq, int ldk, int lds) {
+    void gemm1(T1 *query, const std::tuple<T2 *, int, float *> &keyMat, T3 *score, int M, int N, int headSize, int ldq,
```
Comment (changqi1) [inline src/layers/attention.h:488]: gemm1 -> gemm_query?

File: src/layers/attention.h [hunk 3]
```diff
@@ -504,12 +511,18 @@ class Attention {
 
     // score: M * K(keyLen), value: K * headSize, output: M * headSize
     template <typename T1, typename T2, typename T3>
-    void gemm2(T1 *score, T2 *value, T3 *output, int M, int headSize, int K, int lds, int ldv, int ldo) {
+    void gemm2(T1 *score, const std::tuple<T2 *, int, float *> &valueMat, T3 *output, int M, int headSize, int K,
```
Comment (changqi1) [inline src/layers/attention.h:514]: gemm2 -> gemm_score?

Comment (pujiang2018) [reply src/layers/attention.h:514]: will move to kernel in the progress, will get a better name then.

**Pull request merged**